src/System.Linq/src/System/Linq/Take.SpeedOpt.cs

```diff
@@ -12,5 +13,58 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
+
+        private static IEnumerable<TSource> TakeLastIterator<TSource>(IEnumerable<TSource> source, int count)
+        {
+            Debug.Assert(count > 0);
+
+            if (source is IPartition<TSource> partition)
+            {
+                int length = partition.GetCount(true);
+                if (length > 0)
```

Comment:
It looks like if this is changed to `if (length >= 0)`, then the `else if (length == 0)` below would not be needed.

Reply:
Right. Just will wait for response from https://github.com/dotnet/corefx/pull/36051#discussion_r268010136

Reply:
Done